File: service/src/test/java/hk/qingke/learn/usermanager/service/UserCreateServiceTest.java
package hk.qingke.learn.usermanager.service;

import hk.qingke.learn.usermanager.domain.UserEntity;
import hk.qingke.learn.usermanager.service.exception.UserEmailDuplicationException;
import hk.qingke.learn.usermanager.service.exception.UsernameDuplicateException;
import hk.qingke.learn.usermanager.service.gateway.repository.UserRepository;
import org.apache.commons.lang3.RandomUtils;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.mockito.stubbing.Answer;

import java.util.Optional;

class UserCreateServiceTest {

    private UserCreateService userCreateService;
    private UserRepository userRepository;

    @BeforeEach
    void init() {
        userRepository = Mockito.mock(UserRepository.class);
        userCreateService = new UserCreateService(userRepository);
    }

    @Test
    void create_user_entity() {
        Mockito.when(userRepository.save(Mockito.any(UserEntity.class))).thenAnswer((Answer<UserEntity>) invocationOnMock -> {
            UserEntity userEntity = invocationOnMock.getArgument(0, UserEntity.class);
            userEntity.setId(RandomUtils.nextLong());
            return userEntity;
        });


        UserEntity userEntity = new UserEntity("tom", "1@aB12345678", "test@test.hk");
        userEntity = this.userCreateService.create(userEntity);

        Assertions.assertNotNull(userEntity.getId());
    }

    @Test
    void create_user_entity_fail_given_duplicate_name() {
        Mockito.when(this.userRepository.queryByUsername(Mockito.any(String.class))).thenReturn(Optional.of(new UserEntity()));

        Assertions.assertThrows(UsernameDuplicateException.class, () -> {
            UserEntity userEntity = new UserEntity("tom", "1@aB12345678", "test@test.hk");
            this.userCreateService.create(userEntity);
        });
    }

    @Test
    void create_user_entity_fail_given_duplication_email() {
        Mockito.when(this.userRepository.queryByEmail(Mockito.any(String.class))).thenReturn(Optional.of(new UserEntity()));

        Assertions.assertThrows(UserEmailDuplicationException.class, () -> {
            UserEntity userEntity = new UserEntity("tom", "1@aB12345678", "test@test.hk");
            this.userCreateService.create(userEntity);
        });
    }

}

File: service/src/test/kotlin/hk/qingke/kotlin/usermanager/service/UserCreateServiceTest.kt
package hk.qingke.kotlin.usermanager.service

import hk.qingke.kotlin.usermanager.domain.UserEntity
import hk.qingke.kotlin.usermanager.service.adapter.repository.UserRepository
import hk.qingke.kotlin.usermanager.service.exception.EmailDuplicationException
import hk.qingke.kotlin.usermanager.service.exception.UsernameDuplicationException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.spy
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

internal class UserCreateServiceTest {
    private lateinit var userCreateService: UserCreateService
    private lateinit var userRepository: UserRepository

    @BeforeEach
    fun init() {
        this.userRepository = mock()

        this.userCreateService = spy(UserCreateService(this.userRepository))
    }

    @Test
    fun `user should be created without exception`() {
        whenever(this.userRepository.queryUserByEmail(any())).thenReturn(null)
        whenever(this.userRepository.queryUserByUsername(any())).thenReturn(null)

        val userEntity = UserEntity("tom-test", "12345678@Ab", "tom@test.hk")
        Assertions.assertDoesNotThrow {
            this.userCreateService.create(userEntity)
        }

        verify(this.userRepository).save(userEntity)
        verify(this.userRepository).queryUserByEmail(userEntity.email)
        verify(this.userRepository).queryUserByUsername(userEntity.username)
    }

    @Test
    fun `should throw username duplication exception`() {
        val userEntity = UserEntity("tom-test", "12345678@Ab", "tom@test.hk")

        whenever(this.userRepository.queryUserByEmail(any())).thenReturn(null)
        whenever(this.userRepository.queryUserByUsername(any())).thenReturn(userEntity)

        Assertions.assertThrows(UsernameDuplicationException::class.java) {
            this.userCreateService.create(userEntity)
        }

        verify(this.userRepository).queryUserByUsername(userEntity.username)
    }

    @Test
    fun `should throw email duplication exception`() {
        val userEntity = UserEntity("tom-test", "12345678@Ab", "tom@test.hk")

        whenever(this.userRepository.queryUserByEmail(any())).thenReturn(userEntity)
        whenever(this.userRepository.queryUserByUsername(any())).thenReturn(null)

        Assertions.assertThrows(EmailDuplicationException::class.java) {
            this.userCreateService.create(userEntity)
        }

        verify(this.userRepository).queryUserByEmail(userEntity.email)
    }
}